File: drop_stack_ai/env/merge.cpp
#include "merge.hpp"
#include <deque>
#include <iomanip>
#include <iostream>
#include <string>

/* Pretty-printer (tight ASCII) */
void printBoard(const Board& board)
{
    for (int row = MAX_HEIGHT - 1; row >= 0; --row) {
        for (int col = 0; col < COLUMN_COUNT; ++col) {
            if (row < static_cast<int>(board[col].size()))
                std::cout << std::setw(6) << board[col][row];
            else
                std::cout << std::setw(6) << '.';
        }
        std::cout << '\n';
    }
    std::cout << std::string(COLUMN_COUNT * 6, '-') << '\n';
    for (int col = 0; col < COLUMN_COUNT; ++col)
        std::cout << std::setw(6) << ("C" + std::to_string(col));
    std::cout << "\n\n";
}

/* Drop-and-resolve (bug-free version) */
void dropAndResolve(Board& board, int value, int col)
{
    /* 1.  Drop the new tile */
    board[col].push_back(value);
    int row = static_cast<int>(board[col].size()) - 1;

    /* Remember where that brand-new tile landed */
    const int landingCol = col;
    const int landingRow = row;

    /* 2.  Breadth-first merge resolution */
    std::deque<std::pair<int, int>> fresh{ {col, row} };

    while (!fresh.empty()) {
        auto [c, r] = fresh.front();
        fresh.pop_front();

        if (r >= static_cast<int>(board[c].size())) continue;
        int v = board[c][r];

        /* vertical-down */
        if (r > 0 && board[c][r - 1] == v) {
            board[c][r - 1] *= 2;
            board[c].erase(board[c].begin() + r);
            fresh.push_front({ c, r - 1 });
            if (r < static_cast<int>(board[c].size()))
                fresh.push_back({ c, r });
            continue;
        }

        /* vertical-up */
        if (r + 1 < static_cast<int>(board[c].size()) &&
            board[c][r + 1] == v)
        {
            board[c][r] *= 2;
            board[c].erase(board[c].begin() + r + 1);
            fresh.push_front({ c, r });
            if (r + 1 < static_cast<int>(board[c].size()))
                fresh.push_back({ c, r + 1 });
            continue;
        }

        /* horizontal */
        for (int dc : {-1, 1}) {
            int nc = c + dc;
            if (nc < 0 || nc >= COLUMN_COUNT) continue;
            if (r < static_cast<int>(board[nc].size()) &&
                board[nc][r] == v)
            {
                /* --- only change starts here --- */
                bool keepRight = (r == landingRow && nc == landingCol);
                int  dst = keepRight ? nc : c;   // who keeps the merged tile
                int  src = keepRight ? c : nc;  // who gets erased

                board[dst][r] *= 2;                    // merge
                board[src].erase(board[src].begin() + r);
                /* --- only change ends here --- */

                fresh.push_front({ dst, r });
                if (r < static_cast<int>(board[nc].size()))
                    fresh.push_back({ src, r });
                break;
            }
        }
    }
}

bool gameOver(const Board& board)
{
    for (const auto& col : board)
        if (static_cast<int>(col.size()) >= MAX_HEIGHT)
            return true;
    return false;
}

File: drop_stack_ai/env/merge.py
"""Python implementation of the merge logic for Drop Stack 2048.

This mirrors the behaviour of the C++ code in ``merge.cpp``.  A board is a
list of columns, each column being a list of tile values with the bottom
entry at index ``0``.
"""
from collections import deque
from typing import List


Board = List[List[int]]


def print_board(board: Board, max_height: int | None = None) -> None:
    """Pretty print the board using a tight ASCII layout."""
    column_count = len(board)
    if max_height is None:
        max_height = max((len(col) for col in board), default=0)

    for row in range(max_height - 1, -1, -1):
        line = []
        for col in range(column_count):
            if row < len(board[col]):
                line.append(f"{board[col][row]:6d}")
            else:
                line.append("      .")
        print("".join(line))
    print("-" * (column_count * 6))
    print("".join(f"{f'C{c}':6}" for c in range(column_count)))
    print()


def drop_and_resolve(board: Board, value: int, col: int) -> None:
    """Drop a tile into ``col`` and resolve all resulting merges."""
    # 1. Drop the new tile
    board[col].append(value)
    row = len(board[col]) - 1

    landing_col = col
    landing_row = row

    # 2. Breadth-first merge resolution
    fresh: deque[tuple[int, int]] = deque([(col, row)])

    while fresh:
        c, r = fresh.popleft()

        if r >= len(board[c]):
            continue
        v = board[c][r]

        # vertical-down
        if r > 0 and board[c][r - 1] == v:
            board[c][r - 1] *= 2
            del board[c][r]
            fresh.appendleft((c, r - 1))
            if r < len(board[c]):
                fresh.append((c, r))
            continue

        # vertical-up
        if r + 1 < len(board[c]) and board[c][r + 1] == v:
            board[c][r] *= 2
            del board[c][r + 1]
            fresh.appendleft((c, r))
            if r + 1 < len(board[c]):
                fresh.append((c, r + 1))
            continue

        # horizontal
        for dc in (-1, 1):
            nc = c + dc
            if nc < 0 or nc >= len(board):
                continue
            if r < len(board[nc]) and board[nc][r] == v:
                keep_right = r == landing_row and nc == landing_col
                dst = nc if keep_right else c
                src = c if keep_right else nc

                board[dst][r] *= 2
                del board[src][r]

                fresh.appendleft((dst, r))
                if r < len(board[nc]):
                    fresh.append((src, r))
                break


def game_over(board: Board, max_height: int) -> bool:
    """Return ``True`` if any column reached ``max_height``."""
    return any(len(col) >= max_height for col in board)
